### PR TITLE
DeepTau v2p5 in nanoAOD

### PR DESCRIPTION
**Pull request by mbluj**
#### PR description:
This PR adds working point definitions for a newly integrated DeepTau v2p5. 
A corresponding study with details about WP threshold derivation and tau efficiency/mis-ID rate plots for both Run2 UL and Run 3 samples can be found [here](https://indico.cern.ch/event/1173018/contributions/4926652/attachments/2466518/4237200/wp_eff_final.pdf).

In particular, the changes in this PR are:
* WP thresholds values for DeepTau v2p5 are added to runTauIdMVA.py and thus to miniAOD;
* For nanoAOD, add a dedicated set of variables `_deepTauVars2018v2p5` to `taus_cff.py`,
* Unify WP masking interface to a single function `_tauIdWPMask()`;
* Option to compute WP flags from raw scores (`from_raw` argument in `_tauIdWPMask()`) given the threshold values, instead of reading them directly from MINIAOD;
* Change the format of storing WPs from bitmask to integer values for user-friendliness [see note 1]. 

Backport to 12_4_X foreseen.

[Note 1]: Change of the format of storing WPs affects also WPs of already present tauIDs and thus is not backward compatible. For instance if old versions of nano (with old eras) were produced with this PR integrated they would have format which is not backward compatible with the same versions without this PR. We think, however, that there is not realistic use-case, i.e. old versions of nano are not produced with new releases. Keeping backward compatibility is possible, but implementation will be painful. Comments from experts are welcome.

#### PR validation:

Successfully tested with the "limited" set of matrix tests and a custom nanoAOD production.

### CODE REVIEW

**Comment by gouskos** [comment]
please test

**Comment by mbluj** [comment]
Investigating problem with the failed unit tests.

**Comment by clacaputo** [comment]
@cmsbuild please test

**Comment by mbluj** [comment]
I will be out of office next week (since tomorrow, 15th July) with limited ability to follow this PR. Therefore, my responses on possible comments can be not prompt. Because of the same reason I am going to prepare a backport of this PR to 12_4_X shortly.
FYI, @kandrosov, @ofivite.

**Comment by mariadalfonso** [comment]
is there any difference in mini or mini-se between the 12_5 and 12_4 ?

Reading 12_4 mini and running 12_5 gives some error
https://gitlab.cern.ch/cms-nanoAOD/nanoAOD-integration/-/jobs/23333398#L533

```
----- Begin Fatal Exception 17-Jul-2022 20:24:34 CEST-----------------------
[525](https://gitlab.cern.ch/cms-nanoAOD/nanoAOD-integration/-/jobs/23333398#L525)An exception of category 'Key not found' occurred while
[526](https://gitlab.cern.ch/cms-nanoAOD/nanoAOD-integration/-/jobs/23333398#L526)   [0] Processing  Event run: 1 lumi: 19 event: 1801 stream: 1
[527](https://gitlab.cern.ch/cms-nanoAOD/nanoAOD-integration/-/jobs/23333398#L527)   [1] Running path 'dqmoffline_step'
[528](https://gitlab.cern.ch/cms-nanoAOD/nanoAOD-integration/-/jobs/23333398#L528)   [2] Prefetching for module NanoAODDQM/'nanoDQMMC'
[529](https://gitlab.cern.ch/cms-nanoAOD/nanoAOD-integration/-/jobs/23333398#L529)   [3] Prefetching for module BTagSFProducer/'btagWeightTable'
[530](https://gitlab.cern.ch/cms-nanoAOD/nanoAOD-integration/-/jobs/23333398#L530)   [4] Prefetching for module PATObjectCrossLinker/'linkedObjects'
[531](https://gitlab.cern.ch/cms-nanoAOD/nanoAOD-integration/-/jobs/23333398#L531)   [5] Calling method for module PATTauRefSelector/'finalTaus'
[532](https://gitlab.cern.ch/cms-nanoAOD/nanoAOD-integration/-/jobs/23333398#L532)Exception Message:
[533](https://gitlab.cern.ch/cms-nanoAOD/nanoAOD-integration/-/jobs/23333398#L533)pat::Tau: the ID byVVVLooseDeepTau2018v2p5VSjet can't be found in this pat::Tau.
[534](https://gitlab.cern.ch/cms-nanoAOD/nanoAOD-integration/-/jobs/23333398#L534)The available IDs are: 'againstElectronDeadECAL' 'decayModeFinding' 'decayModeFindingNewDMs' 'byCombinedIsolationDeltaBetaCorrRaw3Hits' 'byLooseCombinedIsolationDeltaBetaCorr3Hits' 'byMediumCombinedIsolationDeltaBetaCorr3Hits' 'byPhotonPtSumOutsideSignalCone' 'byTightCombinedIsolationDeltaBetaCorr3Hits' 'chargedIsoPtSum' 'footprintCorrection' 'neutralIsoPtSum' 'neutralIsoPtSumWeight' 'photonPtSumOutsideSignalCone' 'puCorrPtSum' 'chargedIsoPtSumdR03' 'footprintCorrectiondR03' 'neutralIsoPtSumWeightdR03' 'neutralIsoPtSumdR03' 'photonPtSumOutsideSignalConedR03' 'againstMuonLoose3' 'againstMuonTight3' 'byDeepTau2017v2p1VSeraw' 'byLooseDeepTau2017v2p1VSe' 'byMediumDeepTau2017v2p1VSe' 'byTightDeepTau2017v2p1VSe' 'byVLooseDeepTau2017v2p1VSe' 'byVTightDeepTau2017v2p1VSe' 'byVVLooseDeepTau2017v2p1VSe' 'byVVTightDeepTau2017v2p1VSe' 'byVVVLooseDeepTau2017v2p1VSe' 'byDeepTau2017v2p1VSjetraw' 'byLooseDeepTau2017v2p1VSjet' 'byMediumDeepTau2017v2p1VSjet' 'byTightDeepTau2017v2p1VSjet' 'byVLooseDeepTau2017v2p1VSjet' 'byVTightDeepTau2017v2p1VSjet' 'byVVLooseDeepTau2017v2p1VSjet' 'byVVTightDeepTau2017v2p1VSjet' 'byVVVLooseDeepTau2017v2p1VSjet' 'byDeepTau2017v2p1VSmuraw' 'byLooseDeepTau2017v2p1VSmu' 'byMediumDeepTau2017v2p1VSmu' 'byTightDeepTau2017v2p1VSmu' 'byVLooseDeepTau2017v2p1VSmu' 'byDeepTau2018v2p5VSeraw' 'byDeepTau2018v2p5VSjetraw' 'byDeepTau2018v2p5VSmuraw' .
[535](https://gitlab.cern.ch/cms-nanoAOD/nanoAOD-integration/-/jobs/23333398#L535)
```

**Comment by kandrosov** [comment]
> is there any difference in mini or mini-sw between the 12_5 and 12_4 ?

Yes, in 12_4 miniAOD DeepTau v2p5 WP are not stored, while for future productions (e.g. re-miniAOD) they will be included.
Therefore, for nano production in 12_4, WPs are recomputed from raw by setting `from_raw=True` [here](https://github.com/cms-sw/cmssw/pull/38751/files#diff-d54676262d2e5326ee3455e57747fd476b12be16993a8eb0f4794e8771f6526fR148-R159), while in master `fram_raw=False` (default value) [here](https://github.com/cms-sw/cmssw/pull/38726/files#diff-d54676262d2e5326ee3455e57747fd476b12be16993a8eb0f4794e8771f6526fR148-R156).

**Comment by mbluj** [comment]
@mariadalfonso, recent commit (https://github.com/cms-sw/cmssw/pull/38726/commits/dd403613f92ee9125087af6ca19ac47f30a71bdc) adds a customization function to add items missing in Run-2 UL samples needed by nano v10 (now only deepTau v2p5). To be used as follows:
```
--era Run2_2016,run2_nanoAOD_106Xv2  --customise="PhysicsTools/NanoAOD/nano_cff.nanoAOD_customizeV10"
--era Run2_2016_HIPM,run2_nanoAOD_106Xv2 --customise="PhysicsTools/NanoAOD/nano_cff.nanoAOD_customizeV10"
--era Run2_2017,run2_nanoAOD_106Xv2 --customise="PhysicsTools/NanoAOD/nano_cff.nanoAOD_customizeV10"
--era Run2_2018,run2_nanoAOD_106Xv2 --customise="PhysicsTools/NanoAOD/nano_cff.nanoAOD_customizeV10"
```
It was verified wiht a custom workflow (with a Summer20UL18 input, cmsDriver command below [1]) that the customization behaves as expected.

A counterpart commit for 12_4_X will be prepared shorty.

[1] 
```
cmsDriver.py Summer20UL18NanoAODv10_123X_upgrade2018_realistic_v2 --filein dbs:/GluGluHToTauTau_M125_TuneCP5_13TeV-powheg-pythia8/RunIISummer20UL18MiniAODv2-106X_upgrade2018_realistic_v16_L1v1-v3/MINIAODSIM --fileout file:RunIISummer20UL18NanoAODv10_123X_upgrade2018_realistic_v2.root --mc --eventcontent NANOAODSIM --datatier NANOAODSIM --conditions 123X_upgrade2018_realistic_v2 --step NANO --nThreads 2 --era Run2_2018,run2_nanoAOD_106Xv2 -n 1000  --customise="PhysicsTools/NanoAOD/nano_cff.nanoAOD_customizeV10"
```

**Comment by mariadalfonso** [comment]
please test

**Comment by mariadalfonso** [comment]
> @mariadalfonso, recent commit ([dd40361](https://github.com/cms-sw/cmssw/commit/dd403613f92ee9125087af6ca19ac47f30a71bdc)) adds a customization function to add items missing in Run-2 UL samples needed by nano v10 (now only deepTau v2p5). To be used as follows:


@mbluj we need to re-run the ID also for this modifier, so that we can use the same release for the WinterRun3 EarlyMC
--era Run3,run3_nanoAOD_122 --customise="PhysicsTools/NanoAOD/V10/nano_cff.nanoAOD_customizeV10"

both here and in the Backport PR

**Comment by kandrosov** [comment]
> > @mariadalfonso, recent commit ([dd40361](https://github.com/cms-sw/cmssw/commit/dd403613f92ee9125087af6ca19ac47f30a71bdc)) adds a customization function to add items missing in Run-2 UL samples needed by nano v10 (now only deepTau v2p5). To be used as follows:
> 
> @mbluj we need to re-run the ID also for this modifier, so that we can use the same release for the WinterRun3 EarlyMC --era Run3,run3_nanoAOD_122 --customise="PhysicsTools/NanoAOD/V10/nano_cff.nanoAOD_customizeV10"
> 
> both here and in the Backport PR

@mariadalfonso there are no era-dependent conditions in `nano_cff.nanoAOD_customizeV10`, so it should also work for `run3_nanoAOD_122` if customise is applied. Or I'm missing something?

File: PhysicsTools/NanoAOD/python/nano_cff.py [hunk 1]
```diff
@@ -322,3 +322,29 @@ def nanoWmassGenCustomize(process):
                run2_nanoAOD_106Xv1 )
 _modifiers.toModify(linkedObjects,lowPtElectrons="")
 _modifiers.toModify(simpleCleanerTable,lowPtElectrons="")
+
+###add items missing in Run-2 UL samples needed for V10
+def nanoAOD_customizeV10(process):
```
Comment (mariadalfonso) [inline PhysicsTools/NanoAOD/python/nano_cff.py:327]: we should not use the customise function here but we want to use modifier to switch the mini, and we might need to nest this also in the V11, V12 ...


_modifiers = (run2_nanoAOD_106Xv2 | run3_nanoAOD_122)
(_modifiers) call your function to re-run the ID

File: PhysicsTools/NanoAOD/python/taus_cff.py [hunk 2]
```diff
@@ -179,6 +216,16 @@ def _tauId8WPMask(pattern,doc):
     era.toModify(tauTable.variables,
                  idAntiEleDeadECal = Var("tauID('againstElectronDeadECALForNano')", bool, doc = "Anti-electron dead-ECal discriminator"),
     )
+(run2_nanoAOD_92X | run2_nanoAOD_94XMiniAODv1| run2_nanoAOD_94XMiniAODv2 | \
+ run2_nanoAOD_94X2016 | run2_nanoAOD_102Xv1 | run2_nanoAOD_106Xv1 | \
+ run2_nanoAOD_106Xv2).toModify(tauTable.variables,
```
Comment (mariadalfonso) [inline PhysicsTools/NanoAOD/python/taus_cff.py:221]: rm  run2_nanoAOD_106Xv2 from this list

**Comment by mariadalfonso** [comment]
> > > @mariadalfonso, recent commit ([dd40361](https://github.com/cms-sw/cmssw/commit/dd403613f92ee9125087af6ca19ac47f30a71bdc)) adds a customization function to add items missing in Run-2 UL samples needed by nano v10 (now only deepTau v2p5). To be used as follows:
> > 
> > 
> > @mbluj we need to re-run the ID also for this modifier, so that we can use the same release for the WinterRun3 EarlyMC --era Run3,run3_nanoAOD_122 --customise="PhysicsTools/NanoAOD/V10/nano_cff.nanoAOD_customizeV10"
> > both here and in the Backport PR
> 
> @mariadalfonso there are no era-dependent conditions in `nano_cff.nanoAOD_customizeV10`, so it should also work for `run3_nanoAOD_122` if customise is applied. Or I'm missing something?

do not rely on the customise to with these, but use the modifiers  to switch among the ID

File: PhysicsTools/NanoAOD/python/nano_cff.py (hunk 1 above)
Comment (mbluj) [reply PhysicsTools/NanoAOD/python/nano_cff.py]: DeepTauID v2p5 will be added with `run2_nanoAOD_106Xv2` and `run3_nanoAOD_122` modifiers using existing function to add tauIDs; the customization function `nanoAOD_customizeV10` will be removed.

**Comment by mbluj** [comment]
Commit https://github.com/cms-sw/cmssw/pull/38726/commits/bd03a90a52f7a97594b3cd48eda8a070a0d5f95e adds DeepTauID v2p5 to nanoAOD when the `run2_nanoAOD_106Xv2` or `run3_nanoAOD_122` modifier is used (with an already existing function to add tauIDs),  and removes the `nanoAOD_customizeV10` customization function (as it is not needed longer). 

@mariadalfonso please check if it is desired solution. After a green light from you those changes will be backported to 12_4_X.

PR rebased to 12_5_0_pre4.

**Comment by mariadalfonso** [comment]
looks linear to me now and easy to read
https://github.com/cms-sw/cmssw/pull/38726/files#diff-143388ce2c9c75d53018257e6771622d668ba88c1edb3dd8ddbf423656b06e86R280-R281

**Comment by mariadalfonso** [comment]
please test

**Comment by mbluj** [comment]
In https://github.com/cms-sw/cmssw/pull/38726/commits/d0122b1f7436ac8b0ecc4293b6ab955175183fd9 suffixes are added to tauID modules created for MiniAOD and for NanoAOD. This avoids that modules from one workflow are overriding ones from another workflow when both workflows are run together as in `136.72412`. 
This commit does not solve possible issue that deepTauId is run twice, however I assume that it will not happen in a regular production, e.g. Run-2 UL miniAOD will not be produced together with NanoAOD v10 using CMSSW_12_4_X (or newer).

**Comment by mariadalfonso** [comment]
please test

**Comment by mariadalfonso** [comment]
verified the following:
1) change of the mask to int in the storage
2) add the new ID in miniRun3

3) ID changed in the nano v2.1 to v2.5
4) runs on old mini (winter 12_2 and run2)

3) and 4) verified mode directly with the backport PR

**Comment by mariadalfonso** [comment]
+xpog

**Comment by clacaputo** [comment]
+reconstruction

- reco differences are present only in `tau.tauIDs`  and are in line with the content of this PR

**Comment by qliphy** [comment]
+1

**Pull request merged**